File: code/java/ch06/practice/Challenge01TwoSumThreeWays.java
package ch06.practice;

import java.util.*;

/**
 * Challenge 01: Two Sum — Three Ways
 * ====================================
 * Chapter 6: How Fast Is Your Code?
 *
 * PROBLEM
 * -------
 * Given an array of integers and a target sum, find two indices i and j
 * (i < j) such that nums[i] + nums[j] == target.
 *
 * Implement THREE different approaches:
 *   1. solveBrute: O(n^2) — try every pair
 *   2. solveSort:  O(n log n) — sort + two pointers (track original indices)
 *   3. solveHash:  O(n) — hash map lookup
 *
 * Return {i, j} where i < j, or {-1, -1} if no solution exists.
 * The solve() method should call solveHash (the fastest one).
 *
 * INPUT FORMAT
 * ------------
 * Line 1: space-separated integers (nums)
 * Line 2: target integer
 *
 * OUTPUT FORMAT
 * -------------
 * Print two space-separated integers: i j
 *
 * CONSTRAINTS
 * -----------
 * 2 <= nums.length <= 10^4
 * -10^9 <= nums[i] <= 10^9
 * At most one valid answer exists.
 *
 * EXAMPLES
 * --------
 * Input:       Output:
 * 2 7 11 15    0 1
 * 9
 *
 * Input:       Output:
 * 3 3          0 1
 * 6
 *
 * Input:       Output:
 * 1 2 3        -1 -1
 * 10
 *
 * INSTRUCTIONS
 * ------------
 * Implement all three methods. The main method calls solve() which should
 * use the hash approach. Don't change the main method.
 */
public class Challenge01TwoSumThreeWays {

    /**
     * Brute force: O(n^2) — try every pair.
     */
    public static int[] solveBrute(int[] nums, int target) {
        // TODO: Replace this with your solution
        return new int[]{-1, -1};
    }

    /**
     * Sort + two pointers: O(n log n).
     * Remember to return original indices, not sorted indices!
     */
    public static int[] solveSort(int[] nums, int target) {
        // TODO: Replace this with your solution
        return new int[]{-1, -1};
    }

    /**
     * Hash map: O(n) — the fastest approach.
     */
    public static int[] solveHash(int[] nums, int target) {
        // TODO: Replace this with your solution
        return new int[]{-1, -1};
    }

    /**
     * Default solve uses the hash approach.
     */
    public static int[] solve(int[] nums, int target) {
        return solveHash(nums, target);
    }

    // ── Do not change anything below this line ──────────────────────
    public static void main(String[] args) {
        Scanner sc = new Scanner(System.in);
        int[] nums = Arrays.stream(sc.nextLine().trim().split("\\s+"))
                           .mapToInt(Integer::parseInt).toArray();
        int target = Integer.parseInt(sc.nextLine().trim());
        int[] result = solve(nums, target);
        System.out.println(result[0] + " " + result[1]);
        sc.close();
    }
}

File: code/cpp/ch06/practice/challenge_01_two_sum_three_ways.cpp
/*
 * Challenge 1: Two Sum Three Ways
 * ================================
 * Chapter 6: How Fast Is Your Code?
 *
 * PROBLEM:
 *   Given a vector of integers and a target, return the indices of the
 *   two numbers that add up to target.  If no solution exists, return
 *   {-1, -1}.
 *
 *   Implement THREE approaches:
 *     1. solve_brute — O(n^2) check all pairs
 *     2. solve_sort  — O(n log n) sort + two pointers
 *     3. solve_hash  — O(n) hash map
 *   solve() calls solve_hash (the best one).
 *
 * EXAMPLES:
 *   solve({2, 7, 11, 15}, 9) -> {0, 1}
 *   solve({3, 3}, 6)         -> {0, 1}
 *   solve({1, 2, 3}, 10)     -> {-1, -1}
 *
 * CONSTRAINTS:
 *   - 2 <= nums.size() <= 10^5
 *   - Exactly one solution exists (or none, return {-1, -1})
 *   - Return indices in ascending order
 */

#include <iostream>
#include <vector>
using namespace std;

/**
 * Brute force: check every pair. O(n^2).
 */
vector<int> solve_brute(vector<int>& nums, int target) {
    // TODO: Replace this with your solution
    return {-1, -1};
}

/**
 * Sort + two pointers. O(n log n).
 */
vector<int> solve_sort(vector<int>& nums, int target) {
    // TODO: Replace this with your solution
    return {-1, -1};
}

/**
 * Hash map approach. O(n).
 */
vector<int> solve_hash(vector<int>& nums, int target) {
    // TODO: Replace this with your solution
    return {-1, -1};
}

/**
 * Default: calls the best approach.
 */
vector<int> solve(vector<int>& nums, int target) {
    return solve_hash(nums, target);
}

// -- Do not change anything below this line --------------------------
int main() {
    int n;
    cin >> n;
    vector<int> nums(n);
    for (int i = 0; i < n; i++) cin >> nums[i];
    int target;
    cin >> target;
    vector<int> result = solve(nums, target);
    cout << result[0] << " " << result[1] << endl;
    return 0;
}
